IMP-/Small_StepT.thy

```diff
@@ -0,0 +1,179 @@
+lemma big_to_small_helper: "(c, s) \<Rightarrow> t \<Down> s' \<Longrightarrow> t = Suc t' \<Longrightarrow> (c, s) \<rightarrow>* t' \<down> (SKIP, s')"
+proof (induction arbitrary: t' rule: big_step_t_induct)
+  case (Skip s)
+  then show ?case by simp
+next
+  case (Assign x a s)
+  then have "t' = Suc 0" by auto
+  moreover have "(x ::= a, s) \<rightarrow> (SKIP, s(x := aval a s))" by blast
+  ultimately show ?case by blast
+next
+  case (Seq c1 s1 x s2 c2 y s3 z)
+  then obtain x' y' where suc_ex: "x = Suc x'" "y = Suc y'"
+    by (meson bigstep_progress gr0_implies_Suc)
+  then have "(c1, s1) \<rightarrow>* x' \<down> (SKIP, s2)" using Seq by auto
+  moreover have "(c2, s2) \<rightarrow>* y' \<down> (SKIP, s3)" using Seq suc_ex by auto
+  ultimately have "(c1 ;; c2, s1) \<rightarrow>* (x' + y' + 1) \<down> (SKIP, s3)" using seq_comp by simp  
+  then show ?case using Seq suc_ex by simp
+next
+  case (IfTrue s b c1 x t y c2)
+  then obtain x' where "x = Suc x'" using bigstep_progress gr0_implies_Suc by blast
+  then show ?case using IfTrue by auto
+next
+  case (IfFalse s b c2 x t y c1)
+  then obtain x' where "x = Suc x'" by (meson bigstep_progress gr0_implies_Suc)
+  then show ?case using IfFalse by auto
+next
+  case (WhileFalse s b c)
+  then show ?case by fastforce
+next
+  case (WhileTrue s1 b c x s2 y s3 z)
+  let ?w = "WHILE b \<noteq>0 DO c"
+  obtain x' y' where suc_def:"x =Suc x'" "y = Suc y'"
+    by (meson WhileTrue.hyps(2) WhileTrue.hyps(3) bigstep_progress gr0_implies_Suc)
+  have "(?w, s1) \<rightarrow>*1 \<down> (c ;; ?w, s1)" using WhileTrue by auto
+  moreover have "(c ;; ?w, s1)\<rightarrow>* x' \<down> (SKIP ;; ?w, s2)" 
+    using WhileTrue by (simp add: suc_def star_seq2) 
+  moreover have "(SKIP ;; ?w, s2) \<rightarrow>* 1 \<down>(?w, s2)" by auto
+  moreover have "(?w, s2) \<rightarrow>* y' \<down> (SKIP, s3)" using WhileTrue  \<open>y = Suc y'\<close> by blast
+  ultimately have "(?w, s1) \<rightarrow>* 1 + x' + 1 + y'\<down> (SKIP, s3)" by (meson rel_pow_sum) 
+  moreover have "t' = Suc (Suc (x' + y')) " using WhileTrue suc_def by auto 
+  thus ?case  using calculation by simp
+qed
+
+lemma big_to_small: "(c, s) \<Rightarrow> Suc t \<Down> s' \<Longrightarrow> (c, s) \<rightarrow>* t \<down> (SKIP,s')"
+  using big_to_small_helper by blast
+
+text "from small to big semantics"
+lemma small1_big_continue:
+  "cs \<rightarrow> cs' \<Longrightarrow> cs' \<Rightarrow> t \<Down> cs'' \<Longrightarrow> cs \<Rightarrow> Suc t \<Down> cs''"
+  apply (induction arbitrary: cs'' t rule:small_step.induct )
+  apply auto
+  apply fastforce
+  done
+
+
+lemma small_to_big :
+ "cs \<rightarrow>* t \<down> (SKIP,s') \<Longrightarrow> cs \<Rightarrow> Suc t \<Down> s' "
+  apply(induction t arbitrary:cs s')
+   apply auto[1]
+  using small1_big_continue by blast
+
+text "Equivalence statement"
+lemma equiv_small_big_pair:
+ "(c,s) \<rightarrow>* t \<down> (SKIP,s') \<longleftrightarrow> (c,s) \<Rightarrow> Suc t \<Down> s' "
```

Comment:
can you add a comment why it is off by one? you are counting steps and configurations.

Reply:
done